STM32/I2C/Src/sm4291.c

```diff
@@ -0,0 +1,310 @@
+**
+** @param[in] hi2c Handle to the STM32 I2C peripheral
+** @param[in] crc  Whether to use CRC protection on communications
+** @param[in] press_min Minimum pressure reading of device (unitless)
+** @param[in] press_max Maximum pressure reading of device (unitless)
+**
+** @return A handle for future I2C communications, or null if an error is detected
+*/
+sm4291_i2c_handle_t* sm4291Init(I2C_HandleTypeDef* hi2c, bool crc, double press_min, double press_max) {
+    assert_param(hi2c);
+
+    /* Temporary device handle while we check things */
+    sm4291_i2c_handle_t temp = {hi2c, crc};
+
+    /* Check the serial is non-zero to see if the I2C bus is working */
+    uint32_t serial = 0;
+    int res = sm4291GetSerial(&temp, &serial);
+    if(res == 0 && serial != 0 && serial != 0xFFFFFFFF) {
+        temp.serial = serial;
+
+        /* Pressure transformation */
+        temp.press_scaler = (press_max - press_min) / PRESS_RANGE;
+        temp.press_offset = press_max - temp.press_scaler * (PRESS_RANGE / 2U);
+
+        void* ret = malloc(sizeof(temp));
+
+        if(ret) {
+            ret = memcpy(ret, &temp, sizeof(temp));
+        }
+
+        return (sm4291_i2c_handle_t*) ret;
+    }
+    else {
+        return NULL;
+    }
+}
+
+/*!
+** @brief Get the temperature from the sensor
+*/
+int sm4291GetTemp(sm4291_i2c_handle_t* i2c, double* result) {
+    int16_t temp;
+    if(0 != sm4291ReadReg(i2c, ADDR_DSP_T, (uint16_t*)&temp)) {
+        return -1;
+    };
+
+    *result = temp * TEMP_ADC_SCALAR + TEMP_ADC_OFFSET;
+    return 0;
+}
+
+/*!
+** @brief Get the pressure from the sensor
+*/
+int sm4291GetPressure(sm4291_i2c_handle_t* i2c, double* result) {
+    assert_param(i2c);
+
+    int16_t press;
+    if(0 != sm4291ReadReg(i2c, ADDR_DSP_P, (uint16_t*)&press)) {
+        return -1;
+    };
+
+    *result = press * i2c->press_scaler + i2c->press_offset;
+    return 0;
+}
+
+/*!
+** @brief Get the serial number from the sensor
+*/
+int sm4291GetSerial(sm4291_i2c_handle_t* i2c, uint32_t* result) {
+    uint16_t ser0;
+    uint16_t ser1;
+
+    if(0 != sm4291ReadReg(i2c, ADDR_SER0, &ser0)) {
+        return -1;
+    };
+
+    if(0 != sm4291ReadReg(i2c, ADDR_SER1, &ser1)) {
+        return -2;
+    };
+
+    *result = (((uint32_t)ser1 << 16U) & 0xFFFF0000U) | ((uint32_t)ser0 & 0xFFFFU);
+    return 0;
+}
+
+/*!
+** @brief Get the status register
+*/
+int sm4291GetStatus(sm4291_i2c_handle_t* i2c, uint16_t* status) {
+    return sm4291ReadReg(i2c, ADDR_STATUS, status);
+}
+
+/*!
+** @brief Get the status sync register
+*/
+int sm4291GetStatusSync(sm4291_i2c_handle_t* i2c, uint16_t* status) {
+    return sm4291ReadReg(i2c, ADDR_STATUS_SYNC, status);
+}
+
+/*!
+** @brief Reset the device
+*/
+int sm4291Reset(sm4291_i2c_handle_t* i2c) {
+    return sm4291WriteReg(i2c, ADDR_CMD, CMD_RESET);
+}
+
+/*!
+** @brief Put the device into sleep mode
+**
+** @note The datasheet doesn't specify how t get the device out of sleep mode
```

Comment:
how to*

Reply:
If we don't know how to get the device out of sleep mode is there a reason why we'd want to allow for putting it into sleep mode?

Reply:
Fair question, but I'll leave it in anyway